src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,256 @@
+                            }
+                            else
+                            {
+                                completedTask.Exception.Handle(e =>
+                                {
+                                    // propagate the original exception.
+                                    throw e;
+                                });
+                            }
+                        }
+                        else if (completedTask.Status == TaskStatus.Canceled)
+                        {
+                            // Throw expected OperationCanceledException
+                            completedTask.GetAwaiter().GetResult();
+                        }
+                    }
+                    else
+                    {
+                        tasks.Remove(parallelAwait);
+                    }
+                }
+            }
+
+            throw lastException;
+        }
+
+        private bool IsRetryableException(AggregateException ex)
+        {
+            int statusCode = 0;
+
+            IReadOnlyCollection<Exception> innerExceptions = ex.Flatten().InnerExceptions;
+
+            if (innerExceptions.Count > 0 && innerExceptions.All(ex => ex is RequestFailedException))
+            {
+                statusCode = (innerExceptions.Last() as RequestFailedException).Status;
+            }
+
+            switch (statusCode)
+            {
+                case (int)HttpStatusCode.RequestTimeout:       // 408
+                case HttpStatusCodeRequestThrottled:           // 429
+                case (int)HttpStatusCode.InternalServerError:  // 500
+                case (int)HttpStatusCode.BadGateway:           // 502
+                case (int)HttpStatusCode.ServiceUnavailable:   // 503
+                case (int)HttpStatusCode.GatewayTimeout:       // 504
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        private IEnumerable<ConfigurationClientState> GetPrioritizedConfigurationClientList()
+        {
+            var startIndex = -1;
+            var clients = new List<ConfigurationClientState>();
+            var i = 0;
+
+            foreach (ConfigurationClientState client in _clients)
+            {
+                if (DateTimeOffset.UtcNow >= client.BackoffEndTime)
+                {
+                    clients.Add(client);
+                    if (startIndex == -1)
+                    {
+                        startIndex = i;
+                    }
+                }
+                ++i;
+            }
+
+            // All configuration clients are in the failed state, so we try all clients regardless.
+            if (startIndex == -1)
+            {
+                clients.AddRange(_clients);
+            }
+            // We have put the available configuration clients in the list first, and populating the rest of the clients even though they might be in the failed state.
```

Comment:
Why do we still want to try them even if they are in "back-off" state?

Reply:
Can you please elaborate?

Reply:
My initial implementation was this - 
1. Only use replicas which are known to be in a good state.
2. If all replicas are in the 'back-off' state, then use all of them. 

@jimmyca15 had an opinion that we wouldn't be trying all replicas even though the user provided us when we know few of them are in a 'back-off' state.
i.e. If user provided us 5 replicas, 1,2, and 3 are in the `back-off' state, then as per my initial logic, we would try replicas 4 and 5 and fail the request if 4 and 5 also failed without trying 1,2 and 3 (already known to be in a 'back-off' state).

Reply:
Thanks. What you described is consistent with my understanding of our last discussion.

I know "2" is a little bit odd. We briefly touched on it in our last discussion.  It's mainly because we don't back off if a user provides only 1 endpoint today\*, so the choice of "2" is to be consistent with the existing behavior. We can either 
    a) honor the backoff in the fail-over scenario and leave some inconsistency with the single endpoint scenario, or 
    b) go with "2" for now and fix the back-off issue for all scenarios in the future.

\* For the single endpoint scenario, we actually keep retrying with incremental backoff when a user sets very long refresh expiration (Avani's change). It mainly targets for the push notification scenario. If a user sets very short refresh expiration, like 5 seconds, we just honor it with no backoff. What we should do is to add incremental backoff (up to, eg, 10 minutes) if the refresh keeps failing.

Reply:
As we discussed in the offline meeting, going with the original implementation.